handlers/omni/v2/fileformat/csv/reader.go

```diff
@@ -0,0 +1,156 @@
+package omniv2csv
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/antchfx/xpath"
+	"github.com/jf-tech/go-corelib/caches"
+	"github.com/jf-tech/go-corelib/ios"
+	"github.com/jf-tech/go-corelib/maths"
+
+	"github.com/jf-tech/omniparser/idr"
+)
+
+// ErrInvalidHeader indicates the header of the CSV input is corrupted, mismatched, or simply
+// unreadable. This is a fatal, non-continuable error.
+type ErrInvalidHeader string
+
+func (e ErrInvalidHeader) Error() string { return string(e) }
+
+// IsErrInvalidHeader checks if an err is of ErrInvalidHeader type.
+func IsErrInvalidHeader(err error) bool {
+	switch err.(type) {
+	case ErrInvalidHeader:
+		return true
+	default:
+		return false
+	}
+}
+
+type reader struct {
+	inputName     string
+	decl          *fileDecl
+	xpath         *xpath.Expr
+	r             *ios.LineNumReportingCsvReader
+	headerChecked bool
+}
+
+func (r *reader) Read() (*idr.Node, error) {
+	if !r.headerChecked {
+		err := r.checkHeader()
+		r.headerChecked = true
+		if err != nil {
+			return nil, err
+		}
+	}
+read:
+	record, err := r.r.Read()
+	if err == io.EOF {
+		return nil, io.EOF
+	}
+	if err != nil {
+		return nil, r.FmtErr("failed to fetch record: %s", err.Error())
+	}
+	n := r.recordToNode(record)
+	if r.xpath != nil && !idr.MatchAny(n, r.xpath) {
+		goto read
+	}
+	return n, nil
+}
+
+func (r *reader) checkHeader() error {
+	if r.decl.HeaderRowIndex == nil {
+		_ = r.jumpTo(r.decl.DataRowIndex - 1)
+		return nil
+	}
+	err := r.jumpTo(*r.decl.HeaderRowIndex - 1)
+	if err == io.EOF {
+		return ErrInvalidHeader(r.fmtErrStr("unable to read header: %s", err.Error()))
+	}
+	header, err := r.r.Read()
+	if err != nil {
+		return ErrInvalidHeader(r.fmtErrStr("unable to read header: %s", err.Error()))
+	}
+	if len(header) < len(r.decl.Columns) {
+		return ErrInvalidHeader(r.fmtErrStr(
+			"actual header column size (%d) is less than the size (%d) declared in file_declaration.columns in schema",
+			len(header), len(r.decl.Columns)))
+	}
+	for index, column := range r.decl.Columns {
+		if strings.TrimSpace(header[index]) != strings.TrimSpace(column.Name) {
+			return ErrInvalidHeader(r.fmtErrStr(
+				"header column[%d] '%s' does not match declared column name '%s' in schema",
+				index+1, strings.TrimSpace(header[index]), strings.TrimSpace(column.Name)))
+		}
+	}
+	_ = r.jumpTo(r.decl.DataRowIndex - 1)
```

Comment:
Why we ignore `io.EOF` here? I know you read again later, but looks handling here is better?

Reply:
changed.